internal/controller/runtime/api_discovery.go

```diff
@@ -0,0 +1,246 @@
+// Copyright 2023 - 2024 Crunchy Data Solutions, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package runtime
+
+import (
+	"context"
+	"errors"
+	"sync"
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/rest"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+
+	"github.com/crunchydata/postgres-operator/internal/logging"
+)
+
+// API is a combination of Group, Version, and Kind that can be used to check
+// what is available in the Kubernetes API. There are four ways to populate it:
+//  1. Group without Version nor Kind means any resource in that Group.
+//  2. Group with Version but no Kind means any resource in that GV.
+//  3. Group with Kind but no Version means that Kind in any Version of the Group.
+//  4. Group with Version and Kind means that exact GVK.
+type API = schema.GroupVersionKind
+
+type APIs interface {
+	Has(API) bool
+	HasAll(...API) bool
+	HasOne(...API) bool
+}
+
+// APISet implements APIs using empty struct for minimal memory consumption.
+type APISet map[API]struct{}
+
+// APISet implmenents [APIs].
+var _ APIs = (*APISet)(nil)
```

Comment:
Okay, after some reading I think I'm tracking, but let me write my understanding for confirmation...

Generally, the line is ensuring that `APISet` type satisfies the `APIs` interface. A blank identifier is used since the variable is only declared for the aforementioned purpose. The code on the right side of the `=` is a type conversion where you are converting a `nil` value to the "pointer to `APISet`" type. I presume the reason for doing this rather than just using `&APISet{}` is to save memory?

Reply:
I don't know why I use `(*T)(nil)` instead of `&T{}` or `new(T)`. This article uses `nil`: https://go.dev/doc/effective_go#blank

🤔 Maybe this belongs inside a test. It's already covered by `TestAPIContext`, so maybe we drop it here.

Reply:
I've pushed two commits that move these into a test. Increases coverage a bit, too.